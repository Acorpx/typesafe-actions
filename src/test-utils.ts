import { createStandardAction, createAsyncAction } from './';

/**
 * Fixtures
 */

export type User = { firstName: string; lastName: string };

export namespace types {
  // Symbol
  export const WITH_SYMBOL_TYPE = (Symbol(1) as any) as 'WITH_SYMBOL_TYPE';
  // String Literal
  export const VERY_DEEP_WITH_TYPE_ONLY = 'VERY_DEEP_WITH_TYPE_ONLY';
  export const WITH_TYPE_ONLY = 'WITH_TYPE_ONLY';
  export const WITH_PAYLOAD = 'WITH_PAYLOAD';
  export const WITH_PAYLOAD_META = 'WITH_PAYLOAD_META';
  export const WITH_MAPPED_PAYLOAD = 'WITH_MAPPED_PAYLOAD';
  export const WITH_MAPPED_PAYLOAD_META = 'WITH_MAPPED_PAYLOAD_META';
}

export const actions = {
  withSymbolType: createStandardAction(types.WITH_SYMBOL_TYPE)(),
  very: {
    deep: {
      withTypeOnly: createStandardAction(types.VERY_DEEP_WITH_TYPE_ONLY)(),
    },
  },
  withTypeOnly: createStandardAction(types.WITH_TYPE_ONLY)<void>(),
  withPayload: createStandardAction(types.WITH_PAYLOAD)<number>(),
  withPayloadMeta: createStandardAction(types.WITH_PAYLOAD_META)<
    number,
    string
  >(),
  withMappedPayload: createStandardAction(types.WITH_MAPPED_PAYLOAD).map(
    (payload: number) => ({ payload })
  ),
  withMappedPayloadMeta: createStandardAction(
    types.WITH_MAPPED_PAYLOAD_META
  ).map((payload: number, meta: string) => ({ payload, meta })),
  asyncAction: createAsyncAction(
    'FETCH_USER_REQUEST',
    'FETCH_USER_SUCCESS',
    'FETCH_USER_FAILURE'
  )<void, User, Error>(),
};

/**
 *  Utils
 */
export function testType<T>(a: T): T {
  return a;
}
