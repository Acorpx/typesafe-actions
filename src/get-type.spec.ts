import { getType } from './';
import { actions, types } from './test-utils';

describe('getType', () => {
  it('with type as symbol', () => {
    const typeLiteral: typeof types.WITH_SYMBOL_TYPE = getType(
      actions.withSymbolType
    );
    expect(typeLiteral).toBe(types.WITH_SYMBOL_TYPE);
    expect(typeLiteral).not.toBe('WITH_SYMBOL_TYPE');
  });

  it('with type as string', () => {
    const typeLiteral: 'WITH_TYPE_ONLY' = getType(actions.withTypeOnly);
    expect(typeLiteral).toBe('WITH_TYPE_ONLY');
  });

  it('with payload', () => {
    const typeLiteral: 'WITH_PAYLOAD' = getType(actions.withPayload);
    expect(typeLiteral).toBe('WITH_PAYLOAD');
  });

  it('with mapped payload', () => {
    const typeLiteral: 'WITH_MAPPED_PAYLOAD' = getType(
      actions.withMappedPayload
    );
    expect(typeLiteral).toBe('WITH_MAPPED_PAYLOAD');
  });
});
