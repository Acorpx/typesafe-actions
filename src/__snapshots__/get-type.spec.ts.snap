// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typeLiteral 1`] = `"\\"WITH_TYPE_ONLY\\""`;

exports[`typeLiteral 2`] = `"\\"WITH_PAYLOAD\\""`;

exports[`typeLiteral 3`] = `"\\"WITH_MAPPED_PAYLOAD\\""`;
